### ISSUE TITLE
Manufacturers distributing without source code -- wireless tri-mode boards, missing functionality such as HE matrix etc.

### ISSUE THREAD

**Issue by tzarc** [qmk/qmk_firmware#24085]
#### This issue is limited to reports of non-compliant vendors -- asking for support or complaining about boards being unavailable (apart from reporting them) will be hidden as off-topic. For maintainers, once a vendor has been added to the list below, please mark the comment as "outdated".

If QMK identifies any licensing violation, all current and future PRs from that manufacturer will be put on hold until sources are provided. QMK's license requires full disclosure of source code for any firmware which is based on QMK, includes any part of QMK, or derives from QMK in any way. This source code disclosure is not negotiable and is automatically agreed to by any designer when they choose to use QMK.

Given the recent ramping-up of Chinese vendors disregarding QMK's license, QMK now actively chooses to prevent merges when violations are found.

Vendors may rectify the situation by providing full source code for their boards, even if it's in their own fork of QMK Firmware.

### Current list of vendors who need to provide source code before any new PRs will be considered are [listed on the license violations page](https://docs.qmk.fm/license_violations#offending-vendors).

---

<details>

<summary>Initial post reporting Royal Kludge boards</summary>

Royal Kludge has issued a bunch of PRs recently with significant ambiguity as to whether or not submitted keyboards are tri-mode wireless or not.
So far there has been insufficient evidence supplied stating "no", they're wired-only boards -- comments on PRs when querying this fact have been ambiguous at best.

Additionally, with the fact that Royal Kludge is currently distributing VIA definitions for tri-mode keyboards based on QMK [here](https://en.rkgaming.com/download/12/) -- including the rk839, also known as RK-R65, which shares the Product ID `0xE453` listed in PR #23559, we can only assume that Royal Kludge is submitting wired board definitions to QMK in order to satisfy VIA's requirement that keyboards must exist in QMK's `master` branch, with no intention to supply source code for wireless boards as per QMK's licensing requirements.

As of the posting of this issue, QMK is putting a hold on all PRs from vendor ID `0x342D` until all sources are made available for the corresponding QMK-based boards listed on the above page (inclusive of wireless bindings where relevant):

* RK61 QMK Wired RGB ANSI (PID `0x6461`)
* R65 QMK/VIA Tri-Mode RGB ANSI (PID `0xE453`)
* R65 QMK/VIA Tri-Mode RGB ISO (PID `0xE47F`)
* R65 QMK/VIA Wired RGB ANSI (PID `0xE453`)
* R65 QMK/VIA Wired RGB ISO (PID `0xE480`)
* R75 QMK/VIA Wired RGB ANSI (PID `0xE484`)
* R75 QMK/VIA Wired RGB ISO (PID `0xE483`)
* R75Pro QMK/VIA Tri-Mode RGB ANSI (PID `0xE487`)
* R75Pro QMK/VIA Tri-Mode RGB ISO (PID `0xE485`)

Additionally, the VIA definitions on the same page also list the following combinations which also need full sources provided, including wireless:

* R75Pro QMK (Tri-mode) - vendor ID `0xBB3F`, product ID `0x0001`
* R75 QMK (Tri-mode) - vendor ID `0x342D`, product ID `0xE482`

![image](https://github.com/qmk/qmk_firmware/assets/2985843/46a1ccea-ac02-426a-97fc-98fdb480be04)
![image](https://github.com/qmk/qmk_firmware/assets/2985843/929ccfba-1f5a-4c12-9c02-1a6b10a9e2f8)

</details>

**Comment by tzarc** [comment]
For clarity, the vendor ID is Westberry Technology, which is also in use by Epomaker, KiiBOOM and others.
There is sufficient evidence that these manufacturers are all subcontracting out to the same third-party firmware provider - @sdk66 seems to be listed as maintainer for all.

Manufacturers other than Royal Kludge are also going to be put on hold for this reason -- the third-party firmware provider must supply source code for all QMK-based boards.

**Comment by sdk66** [comment]
First of all, I must clarify that we do not have the capability to develop a tri-mode QMK solution. For the RK839, RK890, and other RK keyboards, the official tri-mode solutions available are written for mechanical keyboards and not using QMK. We do not possess the ability to create a QMK tri-mode solution. Currently, we are using QMK firmware for wired connections only, and the boards are used in common. If the shared use of the keyboard boards has caused you any inconvenience, I deeply apologize. I wish you a pleasant life. Thank you.

**Comment by tzarc** [comment]
**mmd** added to the list of vendors who now require source code disclosure before any future boards are merged.

Two of the three boards listed on their site (http://www.i-game.tech/qmkjpxl) are tri-mode and are seemingly based on QMK:

![image](https://github.com/user-attachments/assets/45d3ff18-dbfa-4571-9f96-e736f4efb129)

**Comment by drashna** [comment]
Qwertykeys, owlab, neo appear to be selling QMK+via compatible keyboards with tri-mode
https://www.qwertykeys.com/pages/fw

**Comment by tzarc** [comment]
Regarding Qwertykeys, all links to `JSON` are to VIA firmware json definitions, thus either fully or partially based on QMK and thus subject to the QMK license.

![image](https://github.com/user-attachments/assets/3707217b-cae6-4014-8577-028389724a65)
![image](https://github.com/user-attachments/assets/075e972b-ff59-400b-9ecf-17abae4ef657)
![image](https://github.com/user-attachments/assets/2c39ab1d-c674-4a57-a511-e1177247238b)
![image](https://github.com/user-attachments/assets/48951ecd-3cee-4b3d-bc1f-a445e7f1b948)
![image](https://github.com/user-attachments/assets/26230b59-9484-4423-8707-d5f83d4a2311)

**Comment by owlab-git** [comment]
> Qwertykeys, owlab, neo appear to be selling QMK+via compatible keyboards with tri-mode https://www.qwertykeys.com/pages/fw
> 
> on each page with tri-mode wireless keyboards, even with tri-mode selected:
> 
> > All PCBs support VIA and the wired features are based on QMK.

Hi. Owlab has not sold any product that uses qmk-based trimode PCB. The only trimode device, which is voice mini, is bases on a proprietary firmware.
As for Qk and neo products, our trimode solutions are provided by 3rd party solution designers and they are not cooperating at this moment. We will strive to negotiate and see if we can work out a plan to fulfill our responsibility.

**Comment by tzarc** [comment]
BBB Keyboard now on the list, see screenshot on #24131.

**Comment by Epomaker** [comment]
Hi, thanks for the help and tracking. We are actively communicating with the 3rd party vendor for opening source code. According to what they state, they may need around a month to solve this since legal regulation may be involved. We will diligently follow up on this matter and find a way to meet the obligation.

**Comment by tzarc** [comment]
> Hi, thanks for the help and tracking. We are actively communicating with the 3rd party vendor for opening source code. According to what they state, they may need around a month to solve this since legal regulation may be involved. We will diligently follow up on this matter and find a way to meet the obligation.

Thank you for your update!

**Comment by tzarc** [comment]
TKD/Vertex added as they support VIA on their tri-mode Cycle8 board, whilst their [QMK-submitted code](https://github.com/qmk/qmk_firmware/tree/master/keyboards/vertex/cycle8) is wired-only.

VIA support on the tri-mode implies portions of QMK were used, if not the entire firmware.

![image](https://github.com/user-attachments/assets/db2b65a2-33bd-4019-9166-ae27d7977627)

**Comment by tzarc** [comment]
Womier added to the list -- their new RD75:

<img width="790" alt="image" src="https://github.com/user-attachments/assets/791a236a-623f-4dd5-a653-a54c245911bc">

**Comment by mozixun-me** [comment]
https://www.lelelab.work/
https://lelelabworks.feishu.cn/wiki/M4evwGWB4iEfglkT8NSc9y3mnmg
This Chinese keyboard manufacturer, which also seems to be developing based on QMK, has a QMK firmware refresh tool in their guidance documents.

**Comment by AsikaXl** [comment]
Believe me, in the end, it is impossible for them to open source. Their software suppliers, especially for the wireless versions, rely on these codes to make a profit. They will not open source in the foreseeable future.

**Comment by KiiBOOM-Official** [comment]
Hi, thanks for bringing this up. We are currently working on addressing this issue. We also noticed that there could be legal issues involved which might impact the duration of solving the problem. Before being merged into QMK repository, we've added all of the source code we currently have to our repository first in case anyone may need it. Once we have made any progress, we will keep everyone updated.

**Comment by amartyaksinha-UC** [comment]
Shortcut Studio's Bridge75 keyboards (Standard and Plus models) are tri mode PCBs with VIA support; yet I don't find any mention of them in any QMK, VIA or Vial repository (although the vendorID for at least the Plus model's PCB shows up in more than 10 keyboards, and interestingly so does a typo from the provided json file). However, there is no mention of QMK support on [their website](https://shortcutofficial.com/) - they only list VIA compatibility at the very bottom of the page. 

However, the firmware updates (bin files) provided by them use QMK toolbox to flash them.
Moreover, I'm currently facing an [issue with their VIA implementation where keys with modifier keycodes only work in wired mode](https://github.com/the-via/app/issues/262); although I'm not sure if this is a VIA or the underlying firmware's issue.

Given all this, I'd like to know if the Bridge75 models are in the process of obtaining QMK certification; or if they are a part of the manufacturers flouting QMK's license.

**Comment by tzarc** [comment]
Never heard of them, looks like license violations. Even if they’re not based on QMK, the VIA compatibility and likely use of firmware-side code from QMK forces the whole thing to be GPL-licensed.

**Comment by tzarc** [comment]
Royal Kludge RK G68 is a bit of a clusterf*ck. See commentary on #20138 / #24272 / #24274.

Seems as if it was perhaps intentionally submitted as if it were AVR-based but was never intended to be. Photos distinctly show RGB working without being plugged in, as well as keycaps showing Bluetooth profile connectivity -- with subsequent hardware being based on a different architecture.

**Comment by Xelus22** [comment]
https://www.luminkey.com/products/luminkey65-65-layout-hot-swappable-custom-mechanical-keyboard

Luminkey advertises VIA support for their wireless keyboards. 
![image](https://github.com/user-attachments/assets/23ebec33-f85e-46ed-8691-5dba89ed08b0)

![image](https://github.com/user-attachments/assets/b4826c80-8e17-472a-95cd-427a1813eb48)

Luminkey 75 specifically lists QMK/VIAL as this keyboard has both a wired and wireless variant. Where as 65 and 80 are only wireless PCBs available.
![image](https://github.com/user-attachments/assets/1cacb847-c466-4060-b93d-757194c2c0df)

**Comment by Xelus22** [comment]
Zuoya GMK67-S and GMK87 (also known as GMK81) also advertises tri-mode with VIA capability.

It seems Zuoya do not sell directly themselves, but only through other vendors

https://clstech.com.au/products/gmk87-keyboard-kit-white
![image](https://github.com/user-attachments/assets/2a4057aa-9dba-44ee-8c4d-b2283b665612)

https://www.aliexpress.com/item/1005006262276373.html
![image](https://github.com/user-attachments/assets/d785daee-d1ef-46ec-9f10-bda9f6c4c8a4)

https://mecha.store/products/zouya-gmk67-s-barebones-mechanical-keyboard
![image](https://github.com/user-attachments/assets/553eca53-fa30-47cb-a664-5e9411dee968)

**Comment by tzarc** [comment]
OP updated with newer manufacturers.

**Comment by drashna** [comment]
> Does "VIA-compatible" mean the firmware must be QMK thus abiding GPL-3?

This one is complicated.  In theory, they could have properly reverse engineered it.  But if they took via.c, and retooled it to work with whatever firmware, then the code is still covered by GPL (2+) and source code still needs to be produced or to be in voliation.

https://www.gnu.org/licenses/gpl-faq.en.html#GPLRequireSourcePostedPublic

> But if you release the modified version to the public in some way, the GPL requires you to make the modified source code available to the program's users, under the GPL.

In this case, releasing the binaries/firmware is releasing to the public.  And: 

https://www.gnu.org/licenses/gpl-faq.en.html#TranslateCode

> Under copyright law, translation of a work is considered a kind of modification. Therefore, what the GPL says about modified versions applies also to translated versions. The translation is covered by the copyright on the original program.

**Comment by tzarc** [comment]
> But if they took via.c, and retooled it to work with whatever firmware, then the code is still covered by GPL (2+) and source code still needs to be produced or to be in voliation.

For clarity's sake, the modified code isn't the only code that needs to be produced -- the entire firmware is considered a derived work, and thus the source for the entire firmware needs to be produced, even if it's overall not based on QMK. Any use of a modified `via.c` results in this outcome.

**Comment by yulei** [comment]
if the wireless firmware runs on a seperate mcu(do not use any QMK related source codes) and communicated the other mcu (runs QMK based firmware) through peripherals such as spi, uart (like the adafruit uart friends used in many keyboard).  I'd thought in such cases, only released the source codes in the MCU which runs  QMK  should be  enough. Is this right?

**Comment by drashna** [comment]
> if the wireless firmware runs on a seperate mcu(do not use any QMK related source codes) and communicated the other mcu (runs QMK based firmware) through peripherals such as spi, uart (like the adafruit uart friends used in many keyboard). I'd thought in such cases, only released the source codes in the MCU which runs QMK should be enough. Is this right?

Correct.  In the case of via stuff, taking the GPL licensed code, and modifying it to work with a different firmware/code would require disclosure of that source too.

**Comment by KrzysztofPrzygoda** [comment]
> Has there been any consideration for Keychron? It appears that certain (possibly newer) boards are missing published firmware...

It seems that entire [Keychron](https://github.com/qmk/qmk_firmware/tree/master/keyboards/keychron) series are missing:

- [K Pro](https://www.keychron.com/collections/keychron-k-pro-series-keyboard),
- [K Max](https://www.keychron.com/collections/keychron-k-max-series-keyboard),
- [V Max](https://www.keychron.com/collections/keychron-v-max-series-keyboard),
- [Q Max](https://www.keychron.com/collections/keychron-q-max-series-keyboard),
- [Q HE (Magnetic)](https://www.keychron.com/collections/keychron-q-he-series-keyboard),
- [K HE (Magnetic)](https://www.keychron.com/pages/keychron-k2-he-landing-page).

**Comment by ld0614** [comment]
> > Has there been any consideration for Keychron? It appears that certain (possibly newer) boards are missing published firmware...
> 
> It seems that entire [Keychron](https://github.com/qmk/qmk_firmware/tree/master/keyboards/keychron) series are missing:
> 
> * [K Pro](https://www.keychron.com/collections/keychron-k-pro-series-keyboard),
> * [K Max](https://www.keychron.com/collections/keychron-k-max-series-keyboard),
> * [V Max](https://www.keychron.com/collections/keychron-v-max-series-keyboard),
> * [Q Max](https://www.keychron.com/collections/keychron-q-max-series-keyboard),
> * [Q HE (Magnetic)](https://www.keychron.com/collections/keychron-q-he-series-keyboard),
> * [K HE (Magnetic)](https://www.keychron.com/pages/keychron-k2-he-landing-page).

It does appear that at least some of these are in various playground Repos in the QMK fork https://github.com/Keychron/qmk_firmware/tree/wireless_playground/keyboards/keychron which as noted in the original post would satisfy the requirements of the GPL although it would be much nicer if they actually did the PRs to get them upstream

**Comment by adophoxia** [comment]
> although it would be much nicer if they actually did the PRs to get them upstream

Never gonna happen as the code they use to make wireless work is too custom and they've made too many code changes for the QMK team to accept.

**Comment by tzarc** [comment]
Yes, Keychron have been far more receptive than almost all other manufacturers regarding disclosure of sources, and have also been receptive to source requests -- I asked them to fast-track their release for the hall effect keyboard sources from their initially publicised 2024Q4 release and they complied, releasing them last week.

**Comment by JackyJia73** [comment]
https://github.com/qmk/qmk_firmware/pull/23982

**Comment by Xelus22** [comment]
https://www.aliexpress.com/item/1005006440319115.html

Darmoshark also in violation of the QMK source. 

![image](https://github.com/user-attachments/assets/6ac12394-5e44-456f-9174-f99277894de4)

**Comment by badmark** [comment]
Does QMK actually have something in place to enforce license violations? As this thread clearly shows (and this is only a small portion) there are countless keyboards being sold as QMK/VIA keyboards that are very clearly closed source software software pretending to be QMK to interface with VIA, but duping thousands of customer with their false advertising. If QMK started taking these companies to court to not only enforce license violations, but also seek monetary damages, QMK could fund actually expanding and making QMK more versatile, expand compatibility, and establish itself as a trusted name in firmwares. 

I have a stock of keyboards that have QMK printed on the box, not a single one actually has a source code repository, and this is watering down, as well as giving consumers a very bad opinion about QMK.

**Comment by ti-mo** [comment]
> As for Qk and neo products, our trimode solutions are provided by 3rd party solution designers and they are not cooperating at this moment. We will strive to negotiate and see if we can work out a plan to fulfill our responsibility.

@owlab-git Any update on this? I received my Neo Ergo this week and got the trimode pcb because the price difference was negligible. As someone working on OSS full-time, I was pretty disappointed when I found out I need to run closed-source firmware on it. All the more because I supported a business that's violating the GPL of multiple projects that are central to the keyboard hobby. Asking around on Discord, it doesn't look like I'm alone in this.

Is there maybe a workaround in the meantime? Could we build a firmware for a trimode pcb that doesn't contain the wireless software stack? Personally, I don't need the wireless functionality at the moment.

**Comment by badmark** [comment]
I just want to know if folks are aware that some of these VIA keyboards appear to be just closed source software imitating QMK to VIA, as with many of these, if you map a key to "Any" many of the standard QMK codes cannot be entered or passed into QMK.

At least, this has been my experience. 

Again I ask, is someone within the QMK group actually charged with license violations and compliance?

**Comment by tzarc** [comment]
CIDOO: https://drop.com/buy/cidoo-v33-wireless-numpad

**Comment by Xelus22** [comment]
Weikav also in violation.

https://weikav.com/product/weikav-record-alice-v2/

![image](https://github.com/user-attachments/assets/2c5c9878-f064-414c-aab8-a208c7fa415c)
![image](https://github.com/user-attachments/assets/fc51df39-f239-4e32-8373-c1eca05d2ba6)

**Comment by vinorodrigues** [comment]
Add [Chosfox CF81](https://github.com/qmk/qmk_firmware/pull/21437) that slipped in last year.  Clearly advertised as a tri-mode `https: // chosfox.com /products/cf81-pro-prebuilt-keyboard`

**alexotos** just did a review on their Fox65 and that will be the same if that comes up.

**Comment by Cipulot** [comment]
> Add [Chosfox CF81](https://github.com/qmk/qmk_firmware/pull/21437) that slipped in last year. Clearly advertised as a tri-mode `https: // chosfox.com /products/cf81-pro-prebuilt-keyboard`
> 
> **alexotos** just did a review on their Fox65 and that will be the same if that comes up.

In connection to this, `0xFFFE` as a VID seems to be used by multiple keebs that have been merged, not provide full code and do have wireless functionality.

Here are boards that are merged, w/o required code disclosure:
- CHOSFOX CF81 (mentioned by @vinorodrigues )

Then there are muddy cases that I'm not 100% sure about, since there are versions with added letters in the name that are wireless, but may not share the same base FW, like:
- Akko 5108 variants
- Inland V83P (here some online manuals show wireless support with the same name, but there was one occurrence of a case with simple win/mac switch)

Comment for tracking purposes for the time being.

**Comment by zvecr** [comment]
> Add https://github.com/qmk/qmk_firmware/pull/21437 that slipped in last year. Clearly advertised as a tri-mode

<https://switchandclick.com/chosfox-cf81/> suggests a non pro version exists, which is not wireless.

**Comment by vinorodrigues** [comment]
> suggests a non pro version exists

You're correct.

**Comment by The3eard** [comment]
> Weikav also in violation.
> 
> https://weikav.com/product/weikav-record-alice-v2/
> 
> ![image](https://private-user-images.githubusercontent.com/17491233/376187303-2c5c9878-f064-414c-aab8-a208c7fa415c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk1MjQxOTgsIm5iZiI6MTcyOTUyMzg5OCwicGF0aCI6Ii8xNzQ5MTIzMy8zNzYxODczMDMtMmM1Yzk4NzgtZjA2NC00MTRjLWFhYjgtYTIwOGM3ZmE0MTVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDIxVDE1MTgxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk5YTAyMDhlYjY4N2YxM2JjMTFiODdlN2EyODQwYTVhZTMxOTI4ZDc5ODM4MzQxZjliOWZiYmNlM2M1NWY4ZDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.aQZdWBYUI3R5puCFUVgFs4FqBHkJVnXGxpCFwshRsNs) ![image](https://private-user-images.githubusercontent.com/17491233/376187358-fc51df39-f239-4e32-8373-c1eca05d2ba6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk1MjQxOTgsIm5iZiI6MTcyOTUyMzg5OCwicGF0aCI6Ii8xNzQ5MTIzMy8zNzYxODczNTgtZmM1MWRmMzktZjIzOS00ZTMyLTgzNzMtYzFlY2EwNWQyYmE2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDIxVDE1MTgxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZjOGY5MWZjNTg1MmZhYzA3ZTYzNThmZDE2YzRmMGFlMDk0MTRhZWFmOGViYjE2YTNiMWMxMTE3ZDk5NTQ5MTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.syMvpb66691r4pcqCmOxdqwNUnlB1glrNggSl9KWc_I)

We can add Weikav Lucky65

**Comment by juanlufont** [comment]
MyKeyClub / IRISLAB may be another candidate for this list.

I own the "QMK" versions (wired-only ones) of 2 of their boards:

- [JRIS75](https://www.mykeyclub.com/pages/jris75-1)
- [MKC75](https://www.mykeyclub.com/pages/mkc75-preorder-page)

Their code is anywhere to be found in this repository:

https://github.com/qmk/qmk_firmware/tree/master/keyboards/mykeyclub

The only code is for JRIS64, and its [README file](keyboards/mykeyclub/jris65/readme.md) contains an interesting disclaimer:

> In May 2024 Mykeyclub was contacted to see if they had interest in contributing firmware to the QMK project they did not. This code is thus community supported.

I do not expect to see any code coming from MyKeyClub nor IRISLAB anytime soon

**Comment by NoAdO** [comment]
Can't fully understand what happens in this tread, but i've found a json file for via and hex for Royal Kludge R75Pro at https://drive.google.com/drive/folders/1eJe8z69hyyY4kqk9i8OI95fs29y4rHe8 , json file testpassed, hex file i did not touch. Mb this helps.

**Comment by TweetyDaBird** [comment]
Add (Ifkb / Ergokbd) to the list. 

As per https://www.ergokbd.com/products/if-corne-wireless2-4g-hotswap-split-mechanical-keyboard-pre-soldered-acrylic-vial-programmable-corne-crkb they are actively promoting their product with QMK and Vial but blatantly and outright refusing to provide ANY source code despite repeated attempts by some buyers.

**Comment by shortcut-git** [comment]
Here are the source code we received from the solution provider, please kindly review and let us know if it suffice to remove us from the list! thanks in advance.
https://github.com/shortcut-git/qmk_firmware/tree/master/keyboards/shortcut
https://github.com/shortcut-git/qmk_firmware/tree/master/keyboards/linker/wireless

**Comment by N1x0** [comment]
Hi @TweetyDaBird,

FYI the source code for IFKB products is available at https://home.ifkb.tech/ which is also linked on the product page of Ergokbd which you reference. To be honest it's still  far from ideal as I myself was a user who tried to get the source code by contacting the vendor on AliExpress I bought it from and was also stone walled. I only came across this when checking qmk toolbox for HID devices and seeing "IFKB sofle" and after a quick google search I found it.


> Add (Ifkb / Ergokbd) to the list.
> 
> As per https://www.ergokbd.com/products/if-corne-wireless2-4g-hotswap-split-mechanical-keyboard-pre-soldered-acrylic-vial-programmable-corne-crkb they are actively promoting their product with QMK and Vial but blatantly and outright refusing to provide ANY source code despite repeated attempts by some buyers.

**Comment by tzarc** [comment]
Incorrect, the sources don't contain the wireless code. The vendor stays for this reason.

**Comment by TweetyDaBird** [comment]
I would have been very happy to be proven wrong, but no. No change. No source provided.

**Comment by N1x0** [comment]
I'm very new to all the QMK stuff so sorry for me asking but when you say there is no wireless code what are you referring to? I thought that all the code on the user manual page  https://home.ifkb.tech/products/if-corne/user-manual would include everything needed.

**Comment by TweetyDaBird** [comment]
That code is the QMK/vial code for the MCU running the receiver. Not the code used to interface with QMK and running the wireless communication. As per the license that has to be supplied as well. Or at least a functional equivalent supplied if not the exact code.

**Comment by tzarc** [comment]
> That code is the QMK/vial code for the MCU running the receiver. Not the code used to interface with QMK and running the wireless communication. As per the license that has to be supplied as well. Or at least a functional equivalent supplied if not the exact code.

Only if it’s running QMK, or includes portions of QMK code. There’s some further investigation required for those boards as to whether or not they’re in breach.

**Comment by TweetyDaBird** [comment]
Yeah, very true, but at the very least it would be running code derivative of QMKs code as it's interfacing with the QMK MCU as a slave. (Although, I'm basically guessing it's using code similar to Mitosis, which I think this is based off).

**Comment by 0xC0FFEEEE** [comment]
> Here are the source code we received from the solution provider, please kindly review and let us know if it suffice to remove us from the list! thanks in advance. https://github.com/shortcut-git/qmk_firmware/tree/master/keyboards/shortcut https://github.com/shortcut-git/qmk_firmware/tree/master/keyboards/linker/wireless

You should follow the [PR checklist](https://docs.qmk.fm/pr_checklist) and submit a PR to this repo for review.

**Comment by tzarc** [comment]
> Here are the source code we received from the solution provider, please kindly review and let us know if it suffice to remove us from the list! thanks in advance. [shortcut-git/qmk_firmware@`master`/keyboards/shortcut](https://github.com/shortcut-git/qmk_firmware/tree/master/keyboards/shortcut?rgh-link-date=2024-12-19T07%3A55%3A36.000Z) [shortcut-git/qmk_firmware@`master`/keyboards/linker/wireless](https://github.com/shortcut-git/qmk_firmware/tree/master/keyboards/linker/wireless?rgh-link-date=2024-12-19T07%3A55%3A36.000Z)

Your site seems to indicate that the HE version also has VIA support, but the repository does not include any HE code as far as I can tell.

**Comment by HorrorTroll** [comment]
> Your site seems to indicate that the HE version also has VIA support, but the repository does not include any HE code as far as I can tell.

Their HE version didn't support VIA, it was using their own software for it. Only the normal and plus version using VIA anyway, which they was provide their code already.

**Comment by tzarc** [comment]
I'll remove them for now but I'd like confirmation from someone with a keyboard at some point.

**Comment by birdspider** [comment]
While writing this [reddit reply](https://www.reddit.com/r/linux_gaming/comments/1ivjtro/comment/meadplu/), I seem to have stumbled over the vendor "Ajazz" (keyboard "Ajazz ak680 max") who seems not confirm with the QMK licence. 

The vendor further provides (as per reddit OP) a web-app for flashing (win/mac only - which prompted the question on reddit) - and has a chinese sounding ssl cert. Also it uses the name qmk it its domain ("app.qmk.top"), which makes it sound "official" and I doubt there is a connections to the qmk project.

I did not see the vendor listed in the license violations table.

**Comment by cachiusa** [comment]
Found this chinese vendor Riccks selling tri-mode keyboard RS8
http://www.riccks.com/col.jsp?id=110
Their website provides via.json, their own qmk_toolbox, binary firmware for that board, but no source provided

**Comment by cachiusa** [comment]
@Gadgetoid Any connection between these vendors?

**Comment by cachiusa** [comment]
> and binds 100% of keyboard functionality to a closed-source binary blob

@Gadgetoid So they can get away with using whatever "QMK" they claim to be using?

**Comment by plastikfan** [comment]
I would like to say that firstly I appreciate the work that the QMK team of volunteer contributors do to provide this software to the mechanical keyboard community.

I am new to this scene and have had a tough time learning the finer nuances. QMK/VIA compatibility is a sensitive topic that I as a newcomer had no idea of, but came to discover during the course of research. It pains me to discover there is a large rogue's list of manufacturers that do not contribute financially to this great project and also violate the terms of use, particularly when most of these firms are the ones newcomers are most likely to come across and buy from without knowing all the finer details and ramifications. As a Wobkey customer, I am firmly within this group, but I bought it in good faith.

Please, QMK/VIA, save your ire for the manufacturers and not the end customer. I feel like a portion of the money I paid for my keyboard should be supporting QMK/VIA. I suppose I could donate directly myself, but having read there is next to no incentive for QMK to support customers with keyboards on the rogue's list, I feel like making a donation is a dead gesture. The onus should be on the manufacturers, that is the only way this would make sense.

If I were to make a new purchase of a mech keyboard, I would probably make a different choice. Perhaps a Keychron would be a better choice since I notice they are not on the list.

FInally, speaking for Wobkey, they actually outsource their software provision to a third company (who, I'm not sure). Wobkey have brought up this issue with them, but their hands are tied. I just wonder why Wobkey would do business with such a company.

**Comment by kbrock84** [comment]
I'm a bit surprised that Keychron is not on the list of offenders considering it has many `Pro` and `Max` variants promoted as QMK wireless, but I see no sources for those.

**Comment by Cipulot** [comment]
> I'm a bit surprised that Keychron is not on the list of offenders considering it has many `Pro` and `Max` variants promoted as QMK wireless, but I see no sources for those.

@kbrock84 Some Keychron boards are not merged into the main QMK repository yet. Th y do maintain th or own separate fork where they publish those boards.
The reason for lack of main merge is the heavy changes applied to some core files, formatting and code flow that would render merging them as of now unfeasible.

GitHub: https://github.com/Keychron/qmk_firmware/tree/playground

They also have different branches like th wireless playground to look at.

**Comment by justMaku** [comment]
SNK is also in violation:
https://www.whatgeek.com/products/skn-bluebird75-tri-mode-wireless-mechanical-keyboard

**Comment by HorrorTroll** [comment]
> SNK is also in violation: https://www.whatgeek.com/products/skn-bluebird75-tri-mode-wireless-mechanical-keyboard

Umm, I don't know what to say. Do they just subsidiary of Keychron or they just asking Keychron make keyboard for them? Because the launcher site here is the same at Keychron launcher.

![Image](https://github.com/user-attachments/assets/6ff212e9-19f1-4bb0-b9c2-c4c1e5b95c7f)

**Comment by justMaku** [comment]
> > SNK is also in violation: https://www.whatgeek.com/products/skn-bluebird75-tri-mode-wireless-mechanical-keyboard
> 
> Umm, I don't know what to say. Do they just subsidiary of Keychron or they just asking Keychron make keyboard for them? Because the launcher site here is the same at Keychron launcher.
> 
> ![Image](https://github.com/user-attachments/assets/6ff212e9-19f1-4bb0-b9c2-c4c1e5b95c7f)

Or they stole the Keychron FW?

EDIT: Seems like contact email on the website is `xiao@keychron.cn` so maybe they are just Keychron subsidiary?
EDIT2: The keyboard does get detected by the `https://launcher.keychron.com`….

**Comment by theRealCarneiro** [comment]
Is Akko in violation with the 5075B https://en.akkogear.com/product/5075b-via/ (this model is the tri-mode variant of the 5075; a PR exists for the wired version, but not for the tri-mode version)?

**Comment by lesshonor** [comment]
> Is Akko in violation with the 5075B https://en.akkogear.com/product/5075b-via/ (this model is the tri-mode variant of the 5075; a PR exists for the wired version, but not for the tri-mode version)?

That product page states:
> For any inquiries about disclosure, please reach out to the Maintainer or Akko Support team.

GPL does not require a public pull request to a specific repository on GitHub. [Akko is permitted to require customers—"users of the program"—to request the source code for their keyboard firmware from customer support.](https://www.gnu.org/licenses/gpl-faq.en.html#GPLRequireSourcePostedPublic)

If you own this keyboard, you have contacted Akko per their instructions, and they have:
- given you the runaround
- outright refused
- pointed you at things which are not source code (e.g. a precompiled binary, the VIA layout definition JSON, the VIA website)

That would be a violation.

**Comment by theRealCarneiro** [comment]
@lesshonor I just emailed them and they shared a link to the repo with the source code, thanks

**Comment by grous0** [comment]
Hello, what about MCHOSE? I use GX87-V2 keyboard and flashed the QMK with the code they provided 

https://github.com/jonylee1986/qmk_firmware_master/tree/mchose_gx87/keyboards/mchose

The thing is this is old or not maintained some functionality is missing like Fn+B to check battery status and some minor stuff that was available for the "default" firmware.

MCU is also Westberry and i think they provide source code of tri-mode with GPLv2 license.

**Comment by grous0** [comment]
Also i can't compile the provided keyboard code with newest QMK (my fork).

```
arm-none-eabi-gcc (GCC) 9.3.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Compiling: keyboards/mchose/common/wireless/transport.c                                            keyboards/mchose/common/wireless/transport.c: In function 'wls_transport_enable':
keyboards/mchose/common/wireless/transport.c:24:13: error: 'keyboard_protocol' undeclared (first use in this function); did you mean 'keyboard_report'?
   24 |             keyboard_protocol = true; // default with true
      |             ^~~~~~~~~~~~~~~~~
      |             keyboard_report
keyboards/mchose/common/wireless/transport.c:24:13: note: each undeclared identifier is reported only once for each function it appears in
 [ERRORS]
 |
 |
 |
make: *** [builddefs/common_rules.mk:362: .build/obj_mchose_gx87_grous/./keyboards/mchose/common/wireless/transport.o] Error 1
```

**Comment by FallenDeity** [comment]
> Hello, what about MCHOSE? I use GX87-V2 keyboard and flashed the QMK with the code they provided
> 
> https://github.com/jonylee1986/qmk_firmware_master/tree/mchose_gx87/keyboards/mchose
> 
> The thing is this is old or not maintained some functionality is missing like Fn+B to check battery status and some minor stuff that was available for the "default" firmware.
> 
> MCU is also Westberry and i think they provide source code of tri-mode with GPLv2 license.

Hey I am looking to tinker with mchose firmware, altho it does look outdated could you point out which functionalities are currently missing, a detailed list would be nice if you're comfortable you can dm me on discord `asher0738` since it would be off topic from this issue. Also did you take a dump of the default firmware before you flashed I couldn't find the default firmware binaries anywhere and would always like to have the default backed up

**Comment by shodanx2** [comment]
Hi,
1.5 year later, why is there not generic tri-mode support ? 
Surely all of these keyboard are using the same solution or very few different designs ?
And since they all have the same solution, why are they not providing their sources ?
Is this because they are using closed source SDKs from their vendor ? 
Or is there no support because "tri mode' wireless capability is a software mote for the vendors, vendors who are sponsoring QMK ?
As a user I feel stuck in the crossfire.

**Comment by nicola02nb** [comment]
> [@lesshonor](https://github.com/lesshonor) I just emailed them and they shared a link to the repo with the source code, thanks

Do you still have that link?

**Comment by cachiusa** [comment]
Seems like Riccks is associated with these 2 brands, same keyboard (RS8) but different name, none of the websites provide anything other than via.json

https://www.elecfox.co/products/linky-87-tkl-customized-mechanical-keyboard-87-keys-wired-wireless-bluetooth-led-gaming-keyboard
https://pmolab.cn/col.jsp?id=113

Update: I sent email requesting source code but got no response

**Comment by damex** [comment]
hi,

is it for manufacturers doing pull requests or it affects regular people too?

can i just go buy some keyboard of my choice, check hardware/pinout and implement it here? if it matches what qmk supports.

i know that weikav is on the list but i did really like their nut65 board and ready to work on implementation. hard to pick anything since literally everything is on violation list. hence the question

**Comment by fcannini** [comment]
Hi there!

I have a [zuoya gmk70](https://www.amazon.com/ZUOYA-GMK70-Split-Mechanical-Keyboard/dp/B0DWJTMYRD) which I've managed to track down **A** source code for it (I'll leave to the jury decide whether this is **THE** source code hehe).
First on [this link](https://www.mathewkb.com/china/gmk70-json/) which led me to [this fork](https://github.com/yangzheng20003/qmk_westberry/tree/tri-mode/keyboards/westberry) and [this fork](https://github.com/WestberryTech/qmk_westberry/tree/tri-mode/keyboards/westberry)

Hope this helps to bring official support to it.

**Comment by emgeka** [comment]
OMOTON may be another vendor requiring review for the non-compliant-vendor list.

**Vendor:** OMOTON  
**Keyboard:** MK212 Gaming Keyboard, ISO-DE variant  
**Connectivity:** USB-C, Bluetooth and 2.4 GHz  
**USB identification in wired mode:** VID `0x36B0`, PID `0x3142`  
**Amazon ASIN:** `B0GF7WCGT7`

The product was supplied through Amazon Vine under an Amazon listing that described it as **"QMK VIA Custom"**. OMOTON also publishes a file named `MK212 Gaming Keyboard.JSON` for VIA on its official support/download page.

The vendor-provided VIA definition works with the keyboard through Raw HID. Testing identified interface 1, usage `0x0061`, usage page `0xFF60`, and working VIA custom-value commands for the main RGB zone and a separate accent-light zone.

OMOTON was asked for the complete corresponding firmware source, including the MK212 board files, MCU and bootloader configuration, matrix and RGB mappings, wireless functionality, and build instructions. Customer service initially redirected the request to the VIA JSON file. After the difference between a VIA definition and firmware source was explained, OMOTON stated that it could not provide material not already published on its website and offered a partial refund. Following a final request, OMOTON replied that it had consulted its development team and reconfirmed that it could not provide the requested firmware information. It offered what it called a full refund instead.

OMOTON did not state that the firmware was independently developed, and it did not deny using QMK or firmware-side VIA code. No complete or functionally equivalent MK212 firmware source has been located.

Could the QMK maintainers review whether this evidence is sufficient to add OMOTON to the non-compliant-vendor list? Sanitised copies of the correspondence and the Amazon Vine order description can be provided if required. No unredacted order or account information will be posted publicly.

References:

- OMOTON product listing: https://www.amazon.de/dp/B0GF7WCGT7
- OMOTON support/download page: https://omoton.com/
- QMK licence-violation guidance: https://github.com/qmk/qmk_firmware/blob/master/docs/license_violations.md

**Comment by ForsakenRei** [comment]
If Qwertykeys and Owlabs are listed separately, maybe worth adding Neo Studio as well (both Owlabs and Neo Studio are ownded by qwertykesy). 
I cannot speak for other newer Neo boards but I have a Neo65 (wired ver, there is a tri-mode ver as well) and I got it working with QMK to flash my custom firmware without issue. But none of the keyboards are in upstream and what they provided is just the firmware, they are not opening any PRs though.

Keyboard list here: https://www.qwertykeys.com/collections/neo-product-collection?srsltid=AU7gw4WM1R-czb9u_prTGI55MbL1DTniMYt5Lp6kch5_uwIooq0Y9gXs
and several of them are listed as VIA support. 

For Neo65 the build guide also instructed you to use QMK toolbox. https://qwertykeys.notion.site/Neo65-Build-Guide-240f14127a97483baa6968a84751c4fa
And in their download center https://www.qwertykeys.com/pages/fw several of the Neo boards can be flashed with QMK toolbox (beta).

**Comment by shodanx2** [comment]
Is there any effort to just wipe all these keyboard of they crap firmware and write some generic firmware for all of them ?
Like the IOT / Tasmotta/ Cloudcutt/ ESPHOME/ BTHOME  etcc.. etc.. have been doing for ages ?
I mean, yes, these manufacturer deserve the hammer for their shear laziness and greed
And they're doing that because they don't know how to do QMK+low power+ tri mode wireless

But come on, this has been a solved problem for years and years, what is the holdup here ? 
Some crappy SDK to program the MCU is not released ?  Really, is that all it is ?  
I find that hard to believe to be frank !
